src/librbd/mirror/snapshot/CreatePrimaryRequest.cc

```diff
@@ -209,7 +209,7 @@ void CreatePrimaryRequest<I>::unlink_peer() {
       // or if it's not linked with any peer (happens if mirroring is enabled
       // on a pool with no peers configured or if UnlinkPeerRequest gets
       // interrupted)
-      if (info->mirror_peer_uuids.size() == 0) {
+      if (info->mirror_peer_uuids.size() == 0 || !info->complete) {
```

Comment:
This is going to need a new test case in `src/test/librbd/mirror/snapshot/test_mock_CreatePrimaryRequest.cc`.

Reply:
And it looks like existing test cases would need to be tweaked too since they don't bother marking snapshots that they inject complete:
```
[ RUN      ] TestMockMirrorSnapshotCreatePrimaryRequest.Success
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/librbd/mirror/snapshot/test_mock_CreatePrimaryRequest.cc: In function 'static librbd::mirror::snapshot::UnlinkPeerRequest<librbd::(anonymous namespace)::MockTestImageCtx> *librbd::mirror::snapshot::UnlinkPeerRequest<librbd::(anonymous namespace)::MockTestImageCtx>::create(librbd::(anonymous namespace)::MockTestImageCtx *, uint64_t, const std::string &, Context *)' thread 7fa819638640 time 2023-03-06T10:27:43.221164+0000
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/librbd/mirror/snapshot/test_mock_CreatePrimaryRequest.cc: 67: FAILED ceph_assert(s_instance != nullptr)
 ceph version Development (no_version) reef (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x1b4) [0x7fa84880a204]
 2: /home/jenkins-build/build/workspace/ceph-pull-requests/build/lib/libceph-common.so.2(+0x143104f) [0x7fa84880a04f]
 3: unittest_librbd(+0x3d19127) [0x55a7610c2127]
 4: unittest_librbd(+0x3d18f4f) [0x55a7610c1f4f]
 5: unittest_librbd(+0x3d1878b) [0x55a7610c178b]
 6: unittest_librbd(+0x3d1868e) [0x55a7610c168e]
 7: unittest_librbd(+0x3d1a0a6) [0x55a7610c30a6]
 8: (Context::complete(int)+0x1f) [0x55a75fc3d53f]
 9: (librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}::operator()() const+0x24) [0x55a75fcfc634]
 10: (boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>::operator()()+0x15) [0x55a75fcfc605]
 11: (void boost::asio::asio_handler_invoke<boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}> >(boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>&, ...)+0x15) [0x55a75fcfc5e5]
 12: (void boost_asio_handler_invoke_helpers::invoke<boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>, librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>(boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>&, librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}&)+0x2f) [0x55a75fcfc5bf]
 13: (void boost::asio::detail::asio_handler_invoke<boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>, librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>(boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>&, boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>*)+0x1d) [0x55a75fcfc57d]
 14: (void boost_asio_handler_invoke_helpers::invoke<boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>, boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}> >(boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>&, boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>&)+0x2d) [0x55a75fcfc51d]
 15: (void boost::asio::detail::handler_work<boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>, boost::asio::io_context::basic_executor_type<std::allocator<void>, 0ul>, void>::complete<boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}> >(boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>&, boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>&)+0x36) [0x55a75fcfc206]
 16: (boost::asio::detail::completion_handler<boost::asio::detail::binder0<librbd::asio::ContextWQ::queue(Context*, int)::{lambda()#1}>, boost::asio::io_context::basic_executor_type<std::allocator<void>, 0ul> >::do_complete(void*, boost::asio::detail::scheduler_operation*, boost::system::error_code const&, unsigned long)+0x9d) [0x55a75fcfc02d]
 17: (boost::asio::detail::scheduler_operation::complete(void*, boost::system::error_code const&, unsigned long)+0x2e) [0x55a7604eb91e]
 18: (boost::asio::detail::strand_service::do_complete(void*, boost::asio::detail::scheduler_operation*, boost::system::error_code const&, unsigned long)+0x98) [0x55a7610625a8]
 19: (boost::asio::detail::scheduler_operation::complete(void*, boost::system::error_code const&, unsigned long)+0x2e) [0x55a7604eb91e]
 20: (boost::asio::detail::scheduler::do_run_one(boost::asio::detail::conditionally_enabled_mutex::scoped_lock&, boost::asio::detail::scheduler_thread_info&, boost::system::error_code const&)+0x1e8) [0x55a7604f00a8]
 21: (boost::asio::detail::scheduler::run(boost::system::error_code&)+0xf7) [0x55a7604efb97]
 22: (boost::asio::io_context::run()+0x2e) [0x55a761bddc1e]
 23: (ceph::async::io_context_pool::start(short)::{lambda()#1}::operator()() const+0x1c) [0x55a761bddbdc]
 24: (void std::__invoke_impl<void, ceph::async::io_context_pool::start(short)::{lambda()#1}>(std::__invoke_other, ceph::async::io_context_pool::start(short)::{lambda()#1}&&)+0x1d) [0x55a761bddbad]
 25: (std::__invoke_result<ceph::async::io_context_pool::start(short)::{lambda()#1}>::type std::__invoke<ceph::async::io_context_pool::start(short)::{lambda()#1}>(ceph::async::io_context_pool::start(short)::{lambda()#1}&&)+0x1d) [0x55a761bddb7d]
 26: (std::invoke_result<ceph::async::io_context_pool::start(short)::{lambda()#1}>::type std::invoke<ceph::async::io_context_pool::start(short)::{lambda()#1}>(ceph::async::io_context_pool::start(short)::{lambda()#1}&&)+0x1d) [0x55a761bddb4d]
 27: _ZZ17make_named_threadIZN4ceph5async15io_context_pool5startEsEUlvE_JEESt6threadSt17basic_string_viewIcSt11char_traitsIcEEOT_DpOT0_ENKUlSA_SD_E_clIS3_JEEEDaSA_SD_()
 28: _ZSt13__invoke_implIvZ17make_named_threadIZN4ceph5async15io_context_pool5startEsEUlvE_JEESt6threadSt17basic_string_viewIcSt11char_traitsIcEEOT_DpOT0_EUlSB_SE_E_JS4_EESA_St14__invoke_otherOT0_DpOT1_()
 29: _ZSt8__invokeIZ17make_named_threadIZN4ceph5async15io_context_pool5startEsEUlvE_JEESt6threadSt17basic_string_viewIcSt11char_traitsIcEEOT_DpOT0_EUlSB_SE_E_JS4_EENSt15__invoke_resultISA_JDpSC_EE4typeESB_SE_()
 30: _ZNSt6thread8_InvokerISt5tupleIJZ17make_named_threadIZN4ceph5async15io_context_pool5startEsEUlvE_JEES_St17basic_string_viewIcSt11char_traitsIcEEOT_DpOT0_EUlSC_SF_E_S6_EEE9_M_invokeIJLm0ELm1EEEEvSt12_Index_tupleIJXspT_EEE()
 31: _ZNSt6thread8_InvokerISt5tupleIJZ17make_named_threadIZN4ceph5async15io_context_pool5startEsEUlvE_JEES_St17basic_string_viewIcSt11char_traitsIcEEOT_DpOT0_EUlSC_SF_E_S6_EEEclEv()
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/run-rbd-unit-tests.sh: line 20: 114225 Aborted                 RBD_FEATURES=$i unittest_librbd
```

Reply:
@idryomov how do I run this single test?

```
[pkalever@localhost build]$ ctest -R "TestMockMirrorSnapshotCreatePrimaryRequest"
Test project /home/pkalever/ceph/build
No tests were found!!!
```

Reply:
RBD_FEATURES=61 ./bin/unittest_librbd --gtest_filter=TestMockMirrorSnapshotCreatePrimaryRequest.Success